File: Python/654. Maximum Binary Tree.py
'''
递归
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        def traversal(nums, begin, end):
            if begin >= end:
                return None
            max_ = nums[begin]
            max_index = begin
            for i in range(begin, end):
                if nums[i] > max_:
                    max_ = nums[i]
                    max_index = i
            node = TreeNode(max_)
            node.left = traversal(nums, begin, max_index)
            node.right = traversal(nums, max_index+1, end)
            return node
        return traversal(nums, 0, len(nums))

File: C++/654. Maximum Binary Tree.cpp
/*
ตÝน้
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* traversal(vector<int>& nums, int begin, int end) {
        if (begin >= end) return NULL;
        int max = nums[begin];
        int max_index = begin;
        for (int i = begin; i < end; i++) {
            if (nums[i] > max) {
                max = nums[i];
                max_index = i;
            }
        }
        TreeNode* node = new TreeNode(max);
        node->left = traversal(nums, begin, max_index);
        node->right = traversal(nums, max_index + 1, end);
        return node;
    }
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return traversal(nums, 0, nums.size());
    }
};